MoreLinq.Test/ScanByTest.cs

```diff
@@ -0,0 +1,120 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2019 Leandro F. Vieira (leandromoh). All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq.Test
+{
+    using System;
+    using NUnit.Framework;
+
+    [TestFixture]
+    public class ScanByTest
+    {
+        [Test]
+        public void ScanByIsLazy()
+        {
+            new BreakingSequence<string>().ScanBy(
+                BreakingFunc.Of<string, int>(),
+                BreakingFunc.Of<int, char>(),
+                BreakingFunc.Of<char, int, string, char>());
+        }
+
+        [Test]
+        public void ScanBy()
+        {
+            var source = new[]
+            {
+                "ana",
+                "beatriz",
+                "carla",
+                "bob",
+                "davi",
+                "adriano",
+                "angelo",
+                "carlos"
+            };
+
+            var result =
+                from x in
+                    source.ScanBy(
+                        item => item.First(),
+                        key => (Element: default(string), Key: key, State: (int)key - 1),
+                        (state, key, item) => (Element: item, Key: char.ToUpper(key), State: state.State + 1))
+                select (x.Key, (x.Value.Element, x.Value.Key, x.Value.State - x.Key));
```

Comment:
I'm not sure what you intended with the `x.Value.State - x.Key` expression. Removing the subtraction of `x.Key` doesn't change anything. The test still passes so it tells me that it's not serving any purpose. In fact, the whole of the last projection can be removed and so the test code just becomes as simple as:

```c#
            var result =
                source.ScanBy(
                    item => item.First(),
                    key => (Element: default(string), Key: key, State: (int)key - 1),
                    (state, key, item) => (item, char.ToUpper(key), state.State + 1));

            result.AssertSequenceEqual(
               KeyValuePair.Create('a', ("ana",     'A', 0)),
               KeyValuePair.Create('b', ("beatriz", 'B', 0)),
               KeyValuePair.Create('c', ("carla",   'C', 0)),
               KeyValuePair.Create('b', ("bob",     'B', 1)),
               KeyValuePair.Create('d', ("davi",    'D', 0)),
               KeyValuePair.Create('a', ("adriano", 'A', 1)),
               KeyValuePair.Create('a', ("angelo",  'A', 2)),
               KeyValuePair.Create('c', ("carlos",  'C', 1)));
```

Reply:
> @atifaziz without the subtraction on last projection 0 would become 97 ('a'), 1 would be 98 ('b') e so on. Remember that seedSelector now returns the int value for each key as seed.

Sure but my point is that the test passes regardless so it can't be serving any intended purpose.

Reply:
[#f70ee87](https://github.com/morelinq/MoreLINQ/pull/572/commits/f70ee87f2f6ade813668aee8835d3dfa67066455) fix this and the other resquests.

Reply:
[#f70ee87](https://github.com/morelinq/MoreLINQ/pull/572/commits/f70ee87f2f6ade813668aee8835d3dfa67066455) fixed this and the other resquests (removed cast; ToUpperInvariant; removed last projection, unnecessary tuple element names). sorry for the sigle commit, I did not have much free time.